File: server/src/services/transaction/transaction.service.ts
// src/services/transaction/transaction.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Transaction } from '../../models/transaction.entity';
import { CreateTransactionDto } from './dto/create-transaction.dto';

@Injectable()
export class TransactionService {
  constructor(
    @InjectRepository(Transaction)
    private transactionRepo: Repository<Transaction>,
  ) {}

  async findAllByUser(userId: number): Promise<Transaction[]> {
    return this.transactionRepo.find({
      where: { user: { id: userId } },
      relations: ['category', 'user'],
      order: { date: 'DESC' },
    });
  }

  async findOneByUser(id: number, userId: number): Promise<Transaction> {
    const transaction = await this.transactionRepo.findOne({
      where: { id, user: { id: userId } },
      relations: ['category', 'user'],
    });
    if (!transaction) throw new NotFoundException('Transaction not found');
    return transaction;
  }

  async create(userId: number, dto: CreateTransactionDto): Promise<Transaction> {
    const transaction = this.transactionRepo.create({
      amount: dto.amount,
      type: dto.type,
      note: dto.note,
      user: { id: userId } as any,
      category: dto.categoryId ? { id: dto.categoryId } as any : undefined,
    });
    return this.transactionRepo.save(transaction);
  }

  async remove(id: number, userId: number): Promise<void> {
    const transaction = await this.findOneByUser(id, userId);
    await this.transactionRepo.remove(transaction);
  }
}

File: server-java/moneyjar/moneyjar/src/main/java/com/tarungattu/moneyjar/services/transaction/TransactionService.java
package com.tarungattu.moneyjar.services.transaction;

import com.tarungattu.moneyjar.dto.CreateTransactionDto;
import com.tarungattu.moneyjar.dto.UpdateTransactionDto;
import com.tarungattu.moneyjar.models.Category;
import com.tarungattu.moneyjar.models.Transaction;
import com.tarungattu.moneyjar.models.User;
import com.tarungattu.moneyjar.repository.CategoryRepository;
import com.tarungattu.moneyjar.repository.TransactionRepository;
import com.tarungattu.moneyjar.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
public class TransactionService {

    @Autowired
    private TransactionRepository transactionRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    public List<Transaction> findAllByUser(UUID userId) {
        return transactionRepository.findByUserIdOrderByDateDesc(userId);
    }

    public Transaction findOneByUser(UUID id, UUID userId) {
        return transactionRepository.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new RuntimeException("Transaction not found"));
    }

    public Transaction create(UUID userId, CreateTransactionDto dto) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Transaction transaction = new Transaction();
        transaction.setUser(user);
        transaction.setAmount(dto.getAmount());
        transaction.setType(dto.getType());
        transaction.setNote(dto.getNote());

        if (dto.getCategoryId() != null) {
            Category category = categoryRepository.findById(dto.getCategoryId())
                    .orElseThrow(() -> new RuntimeException("Category not found"));
            transaction.setCategory(category);
        }

        return transactionRepository.save(transaction);
    }

    public Transaction updateTransaction(UUID id, UUID userId, UpdateTransactionDto dto) {
        Transaction transaction = findOneByUser(id, userId);

        if (dto.getAmount() != null) transaction.setAmount(dto.getAmount());
        if (dto.getType() != null) transaction.setType(dto.getType());
        if (dto.getNote() != null) transaction.setNote(dto.getNote());
        
        // Handle Category update (nullable or change)
        if (dto.getCategoryId() != null) {
             Category category = categoryRepository.findById(dto.getCategoryId())
                    .orElseThrow(() -> new RuntimeException("Category not found"));
            transaction.setCategory(category);
        }

        return transactionRepository.save(transaction);
    }

    public void remove(UUID id, UUID userId) {
        Transaction transaction = findOneByUser(id, userId);
        transactionRepository.delete(transaction);
    }
}
